dev/dram/device.c

```diff
@@ -93,75 +112,429 @@ void dram_cbt_exit(dram_dev_t *dram,
+                                                dfi_tx_packet_buffer_t *buffer,
+                                                burst_length_t burst_length,
+                                                chipselect_t cs,
+                                                uint16_t rddata_en_offset,
+                                                command_t *command,
+                                                uint16_t time_offset)
+{
+    uint16_t initial_dram_offset = 0;
+    packet_item_t *packet;
+    packet_group_info_t rddata_info;
+
+    // Calculate timestamp of first CA packet (time 0)
+    initial_dram_offset = (buffer->ts_last_packet + time_offset) << dram->cfg->ratio;
+
+    // Calculate rddata offset required for rd packet from time 0
+    // time 0 is identified as first CA packet ts
+    // rddata offset is measured in DRAM CLKs
+    uint8_t rddata_dram_clk = initial_dram_offset + CA_CLK_LEN + dram->cfg->phy_rd_en + rddata_en_offset;
+    uint8_t cycles_per_packet = 1 << dram->cfg->ratio;
+
+    create_packet_group_info(&rddata_info, dram->cfg->ratio, rddata_dram_clk, burst_length);
+
+    create_address_packet_sequence(buffer, dram->cfg->ratio, command, time_offset);
+
+    // Need to add phy_rd_en to get next packet offset.
+    time_offset = rddata_info.dfi_ts_start - buffer->ts_last_packet;
+
+    // If enough delay is required, insert CKE packet to keep clocks cycling.
+    if (time_offset > 1)
+    {
+        create_cke_packet_sequence(buffer, 1);
+        time_offset--;
+    }
+
+    while (rddata_info.phase_remaining)
+    {
+        packet = create_cke_packet_sequence(buffer, time_offset);
+        if (packet == NULL)
+        {
+            return WDDR_ERROR;
+        }
+
+        fill_rddata_packet(packet, &rddata_info, cs, cycles_per_packet);
+
+        time_offset = 1;
+    }
+
+    create_cke_packet_sequence(buffer, 1);
+
+    return WDDR_SUCCESS;
+}
+
+wddr_return_t dram_prepare_rddq_sequence(dram_dev_t *dram,
+                                         dfi_tx_packet_buffer_t *buffer,
+                                         burst_length_t burst_length,
+                                         chipselect_t cs,
+                                         uint16_t rddata_en_offset)
+{
+    command_t command = {0};
+    create_rddq_command(&command, cs);
+    return dram_prepare_read_sequence(dram, buffer, burst_length, cs, rddata_en_offset, &command, 1);
+}
+
+wddr_return_t dram_prepare_rdfifo_sequence(dram_dev_t *dram,
+                                           dfi_tx_packet_buffer_t *buffer,
+                                           burst_length_t burst_length,
+                                           chipselect_t cs,
+                                           uint16_t rddata_en_offset)
+{
+    command_t command = {0};
+    create_rdfifo_command(&command, cs);
+    return dram_prepare_read_sequence(dram, buffer, burst_length, cs, rddata_en_offset, &command, WRITE_LATENCY + BL + TWTR + 1);
```

Comment:
We need to revisit this equation at the end. BL can be replaced with BL_16

Reply:
Is there a better WRITE_LATENCY value we should be using instead of having it hard coded?  Should we add those to the table to have?  Same with tWTR, those are different based on frequency.

Reply:
For documentation, we discussed this. Agreed to keep general and not have offset hardcoded so that other rdfifo sequence timings can be used. This timing will be accounted for at training level such that timing is met correctly when rdfifo is called after wrfifo